src/mlpack/methods/lsh/lshmodel_impl.hpp

```diff
@@ -0,0 +1,607 @@
+    for (size_t j = i + 1; j < numSamples; ++j)
+      distances(d++) = metric::EuclideanDistance::Evaluate(
+          sampleSet.unsafe_col(i), sampleSet.unsafe_col(j));
+  Log::Info << "Computed " << d << " pointwise distances." << std::endl;
+  Timer::Stop("pairwise_distances");
+
+  // Step 3. Estimate statistics of these distances: log(mean(d)), mean(log(d)),
+  // mean(d).
+  distances = arma::pow(distances, 2);
+  this->meanDist = arma::mean(distances);
+  this->logMeanDist = std::log(meanDist);
+  this->meanLogDist = arma::mean(arma::log(distances));
+
+  // Step 4. Select a small part of the sample as 'anchor points'. Use the rest
+  // of the sample as the reference set. Find the k-Nearest Neighbors' distances
+  // from the anchor points for increasing portion of the reference set. Compute
+  // the arithmetic and geometric mean of distances from each anchor to its
+  // k-Nearest Neighbor.
+  // The geometric mean of N numbers is the Nth root of the product of the
+  // numbers. Through logarithmic properties though, this becomes computable
+  // through exponentiating the mean of the logarithms of x:
+  // exp(mean(log(x))) = geometricmean(x).
+
+  // Number of points to use as queries. Use 10% of sample.
+  double anchorsSample = 0.1;
+  size_t numAnchors = (size_t) std::round(anchorsSample * numSamples);
+
+  // Statistics - Arithmetic and geometric means for growing reference set.
+  // Compute one of each for each k.
+  size_t regressionExamples = size_t(
+      std::round((1.0 - anchorsSample) / anchorsSample));
+  arma::mat Ek(regressionExamples, k);
+  arma::mat Gk(regressionExamples, k);
+
+  // For each reference size, calculate the kNN of the anchors. Divide reference
+  // set into equal blocks (block 1 is anchors). In repetition 1, use block 2 as
+  // reference set, in repetition 2, blocks 2 and 3, and so on.
+  size_t refSetStart = numAnchors;
+  size_t refSetEnd = numAnchors;
+  arma::mat queryMat = sampleSet.cols(0, numAnchors - 1);
+  arma::Col<size_t> referenceSizes(regressionExamples);
+
+  Log::Info.ignoreInput = true; // Ignore kNN output.
+  // TODO: Since we've already computed this, avoid calling kNN?
+  for (size_t i = 0; i < regressionExamples; ++i)
+  {
+    refSetEnd += refSetStart - 1;
+
+    cout << "Neighbors "<< refSetStart <<":"<<refSetEnd<<endl;
+
+    // Reference set for kNN
+    arma::mat refMat = sampleSet.cols(refSetStart, refSetEnd);
+    referenceSizes(i) = refMat.n_cols;
+
+    arma::Mat<size_t> neighbors; // Not going to be used but required.
+    arma::mat kNNDistances; // What we need.
+    KNN naive(refMat, true); // true: train and use naive kNN.
+    naive.Search(queryMat, k, neighbors, kNNDistances);
+
+    // Store the squared distances (what we need).
+    kNNDistances = arma::pow(kNNDistances, 2);
+
+    // Compute Arithmetic and Geometric mean of the distances.
+    Ek.row(i) = arma::mean(kNNDistances.t());
+    Gk.row(i) = arma::exp(arma::mean(arma::log(kNNDistances.t()), 0));
```

Comment:
I took a look at the most current code, and I see you are doing `find(kNNDistances > 0)`, but I don't think this will adequately filter duplicates.  Tomorrow I'll try and think about a good way to filter duplicate points; probably the best time to do that is during the calculation of the kNN distances matrix.  (i.e. if we encounter a zero distance, clear the row/column of the matrix and skip to the next point)  I need to think a little bit more about it...